internal/agent/device/device.go

```diff
@@ -111,7 +118,11 @@ func (a *Agent) Run(ctx context.Context) error {
 
 // Stop ensures that the device agent stops reconciling during graceful shutdown.
 func (a *Agent) Stop(ctx context.Context) error {
-	a.cancelFn()
+	a.once.Do(func() {
```

Comment:
note: although the agent is not expected to be thread-safe we need to ensure shutdown is.